### DISCUSSION TITLE
Proper Mo' Creatures (CustomMobSpawner) mod intercompatibility, or implementation

### DISCUSSION THREAD

**Discussion by xJon** [ACGaming/UniversalTweaks#125] [Tweak Suggestions]
[Mo' Creatures](https://github.com/DrZhark/mocreaturesdev/tree/1.12.1/src/main/java/drzhark/mocreatures) uses [CustomMobSpawner](https://github.com/DrZhark/mocreaturesdev/tree/1.12.1/src/main/java/drzhark/customspawner) to spawn its mobs, or at least some of them, sometimes. When launching the game without CustomMobSpawner installed, there's no obvious problem to boot and when you enter a world you will probably even see some Mo' Creatures mobs, but a lot of them will be missing - distinctly the water mobs (also reported over https://github.com/DrZhark/mocreatures/issues/37), and maybe snakes, and other mobs. Also, trying to use Mo' Creatures special in-game config (that was ahead of its time 10 years ago!) [crashes the game](https://mclo.gs/iMIlHk5).

So let's also install CustomMobSpawner for the full Mo' Creatures experience! Right? Well not really, if you (like most others) want to play with other mods and have all of them work as intended; also not if you want vanilla mob farms to work. CustomMobSpawner seems to completely override the entire game's spawning system, so a lot of modded biomes and dimensions will not have any mob spawns in them.

We've already found out that some modded biome support does exist and seems [hardcoded](https://github.com/DrZhark/mocreaturesdev/blob/be53e6257ef0158967c20ed6496bf3cc66a0ad07/src/main/java/drzhark/customspawner/environment/EnvironmentSettings.java#L808), yet that doesn't work for other mods the code doesn't expect. There's the [current tweak](https://github.com/ACGaming/UniversalTweaks/commit/31aea8846a296da78ed7ae163bd612b0bd495886) that adds modded biome support via the manual config, which allows mobs to spawn in those mods' biomes, yet despite spending _hours_ on this, there are some critical issues that tweak doesn't seem to overcome:

1. **No Mystcraft support**
It seems CustomMobSpawner doesn't like Mystcraft's dynamically generating dimensions and usage of biomes. As long as CustomMobSpawner is installed, no kind of mobs will spawn in any Mystcraft age. I was unable to resolve this with the current tweak.
2. **Broken Galacticraft support**
CustomMobSpawner seems to only detect some biomes (e.g. some in Venus, none in the Moon). This means no mobs spawn on the Moon. Perhaps this is due to the multiple biome mod ids packaged together. I was unable to resolve this with the current tweak.
3. **Other issues**
Random, smaller other issues: some mods' mobs will just not spawn, e.g. in Hexxit II I was never able to get a single spawn of Lumen's Embers in the Nether, Project Vibrant Journeys' Starfish on beaches, or even [vanilla bats in caves](https://github.com/xJon/Hexxit-II/issues/162) (despite spending *days* playing with configs to try to resolve this).

All these issues disappear when removing CustomMobSpawner. It's also worth mentioning that enabling `doMobSpawning` in CustomMobSpawner's `Global.cfg` config file does not seem to resolve any of these issues.

Also, this pool of issues is very specific to my use case trying to build modpacks with Mo' Creatures, and I'm sure with the current way CustomMobSpawner works in 1.12 there are plenty more mod incompatibilities, some can be easily found over https://github.com/DrZhark/mocreatures/issues.

**What am I suggesting? Possibly one of the following**:
- Magically making the current tweak fix the critical mod incompatibilities with Galacticraft and Mystcraft (that make the mod quite obsolete in its current state, especially its aquatic creatures).
- New tweak to properly resolve CustomMobSpawner's fussy incompatibilities with other mods' biomes/dimensions (e.g. Mystcraft, Galacticraft), which break _all_ spawns in their biomes/dimensions.
- New tweak to make Mo' Creatures self sustainable without depending on CustomMobSpawner's complete overhauls, so when it's installed on its own all of its creatures/mobs will spawn properly and no features will be lost, whilst having no incompatibilities present.

Last suggestion is particularly a longshot, but considering the mod already semi-works without CustomMobSpawner *might* be a positive indicator? Nevertheless, this is just a suggestion/discussion.
(Also, depending on the scope of any possible solution this might deserve its own mod).

**Comment by xJon** [comment]
Fun fact: without CustomMobSpawner installed, Mo' Creatures' aquatic mobs *do* spawn in this random Mystcraft age, which had an Ocean biome with grass top block.
![javaw_DYngXBunE2](https://user-images.githubusercontent.com/8945347/229661702-d21ec482-20ce-49bc-96e7-6cd16c24343a.png)
Fish grass.

**Comment by ACGaming** [comment]
Closed in favor of the upcoming Mo‘ Creatures fork.
